Document instance migration phases 3 and 4 (#4348)
camunda/camunda-docs: docs/components/concepts/process-instance-migration.md

You can migrate a called process instance in the same way as a regular process instance.

## Dealing with catch events

> An exception to changing the process instance state is specific to catch events.

Reviewer: @korthout moved the first paragraph of the `Jobs, expressions, and input mappings` to the introduction on the top.

This is necessary to ensure that the process instance can be executed according to the new process definition.

> This allows you to add or remove catch events from an active element.

Author: ```suggestion
It allows you to add or remove catch events from an active element.
```

At the same time, you can leave an existing catch event unchanged.
This section explains how to deal with catch events when migrating a process instance.

You decide what happens to the associated event subscription through the mapping instructions for the catch events:

- Migrate catch events: if a catch event is mapped, the associated subscription is migrated.
- Remove catch events: if a catch event in the source process is not mapped, then the associated subscription is closed during migration.
- Add catch events: if a catch event of the target process is not the target of a mapping instruction, then a new subscription is opened during migration.

### Migrate catch events

Catch events can be migrated by providing a mapping instruction between the source and the target catch event.

> Providing a mapping between catch events ensures that the behaviour of the event subscription in the source process is preserved after the migration.

Author: ```suggestion
Providing a mapping between catch events ensures that the event subscription in the source process is preserved after the migration.
```

> This is especially useful when a subscription behaviour is changed in the new version of the process definition, but you want to keep the catch event behaviour of the running processes the same after the migration.

Author: 🔧 I suggest removing this sentence. The sentence below informs us to read the next sections to understand when this is useful. The previous sentences have already provided the information that the event subscription is preserved.

In the following section we will discuss situations where mapping a catch event may or may not be useful.

> #### When to map a catch event?

Author: 🔧 This section seems a bit unnecessary now. Let's remove the heading and perhaps keep the `Let's explain both mapping and non-mapping scenarios with examples.` sentence by moving it to the previous section.

<!--
- Definition of the catch event does not change
  - You want to keep the event subscription the same
  - You want to reset event subscription
- Definition of the catch event changes
  - You want to keep the old event subscription
  - You want to use the new catch event definition to re-initiate the subscription

> -->

Author: 🔧 Let's remove this old comment

Choosing to map the catch event or not gives you control over what happens to the event subscription.
Let's explain both mapping and non-mapping scenarios with examples.

#### A mapping instruction is provided between the catch events

An active user task has been waiting for a timer boundary event.
The timer boundary event is defined as a duration of one week.
The user task has not been completed and has already spent five days waiting for the timer.

![The process instance is waiting at the active user task A with a timer boundary event attached.](assets/process-instance-migration/migration-catch-event-source.png)

> Now we want to [change an inactive part of the process](#changing-the-process-instance-flow-for-inactive-parts) by adding a user task after the user task with the timer boundary event.

Author: 🔧 The image shows that the user task is added after the timer boundary event, instead of after the user task.

```suggestion
Now we want to [change an inactive part of the process](#changing-the-process-instance-flow-for-inactive-parts) by adding a user task after the timer boundary event.
```

When migrating the process instance, we want to keep the timer unchanged.
Instead of waiting for the full time defined by the target process' timer boundary event, we only want to wait for the remaining two days.
To achieve that, you must map the timer boundary event to the timer boundary event in the target process.
This ensures the timer is migrated (_the associated subscription is migrated_) and the duration is preserved.
To achieve this for the example above, the mapping between active user tasks `A` -> `A` and timer boundary events `Timer1` -> `Timer2` must be provided.
Assuming that the timer boundary event is defined as 2 weeks duration in the target process, the process instance will look like following after the migration:

![The process instance is waiting at the active user task A with the migrated timer boundary event attached.](assets/process-instance-migration/migration-catch-event-different-target.png)

Mapping catch events applies to other event types as well.

> For example, if you want to keep the message name the same for a message event subprocess start event, you should map the message start event when migrating the process instance.

Author: 🔧 Just want to be a bit more explicit

```suggestion
For example, if you want to keep the message name the same for a message event subprocess, you should map the start event of the event subprocess when migrating the process instance.
```

> Another example would be to preserve a signal name for an intermediate signal catch event, you should map the signal catch event when to the one in the target while migrating the process instance.

Author: This behavior is specific to event-based gateway. A regular intermediate catch event **must always be mapped** because it's an active element.

```suggestion
Another example would be to preserve a signal name for an intermediate signal catch event attached to an event-based gateway. In this case, you should map the signal catch event to the one in the target while migrating the process instance.
```

You decide what happens to the associated event subscription through the mapping instructions for the catch events:

- Migrate catch events: if a catch event is mapped, the associated subscription is migrated.
- Remove catch events: if a catch event in the source process is not mapped, then the associated subscription is closed during migration.

> - Create catch events: if a catch event of the target process is not the target of a mapping instruction, then a new subscription is opened during migration.

Author: 👍 I like it

> This results in new keys for the service task as well as the job.
> :::
> 
> ## Internal Execution
> 
> In the following example, we will explain the internal execution steps of process instance migration.
> It is recommended that you read these steps to fully understand the power of process instance migration.
> 
> Migration of a process instance consists of the following steps:
> 
> - Validation
> - Migration of mapped active elements
> - Unsubscribing from catch events
> - Subscribing to catch events
> - Migrating catch events subscriptions

Author: 🔧 The steps are actually more involved and we don't migrate all active elements before dealing with the catch events. More correct would be the following.
```suggestion
- Validation
- Migration of process instance
- Migration of global variables
- For each active element
  - Migration of element instance
  - Migration of associated job
  - Migration of associated incident
  - Migration of associated user task
  - Migration of local variables
  - Unsubscribing from catch events
  - Subscribing to catch events
  - Migrating catch events subscriptions
```

💭 However, we should consider whether we want to go into so much detail about the implementation, as it means it becomes harder the change if we have such a need in the future. What does the reader benefit from this knowledge? If we see a need to document this, then I'd suggest a more ambiguous description to allow more flexibility in the future:

```
- Validation
- Migration of process instance, and global variables
- Migration of each active element (including associated jobs, incidents, user tasks, local variables, and event subscriptions)
```

Reviewer: > What does the reader benefit from this knowledge?

Good question. Using migration tool can be complicated especially for bigger process definitions with a lot of active elements and we have a bunch of rejection messages. Explaining the order of execution would be useful to reason about these rejections and their orders. I also agree, we can further simplify this by your later suggestion.

❌ to myself. While overviewing the new page before merge. I realised, I indicated the execution order the other way around. It is actually breadth first. I will fix that one as well.

Reviewer: I chose to opt-out mentioning user tasks because it is just an implementation detail of user tasks.

Note that this results in new keys for the process instance and its associated variables, element instances, and other entities.
:::

> ## Supported BPMN elements

Author: 🔧 IMO, this should be part of the limitations section. While we take the positive approach of documented what is supported, it actually describes that there are elements that cannot be migrated yet.
